### ISSUE TITLE
Converting section style to .css instead of adding "style" to html markup

### ISSUE THREAD

**Issue by vivalafit** [ThemeFuse/Unyson#3085]
Greeting to everyone .
So, i was customizing some framework features . As you can see from title name - i was dealing with section shortcode element.
I faced with some issues and questions  , which i will list right now : 
**_1) When we are adding some styles to section element (for example back-ground color in screenshot below) they are added like `styles` section  in html `div`/`section` blocks , which is not appopriate due to coding standarts and simply - if we have a lot "customizable elements" it will greatly decrease page loading speed , because everytime we refresh page - we have to load `style` again and again ._**

![image](https://user-images.githubusercontent.com/24575491/34082747-ceccc790-e36c-11e7-9f3c-4c63ea523b5b.png)
![image](https://user-images.githubusercontent.com/24575491/34082797-d65efed2-e36d-11e7-88ab-470445c867e0.png)
So , i came up with this solution like this : in `view.php` we just declaring `$_SESSION` variables (that get values from elements from `$atts`  array ) which will be a part of `css` code and drag them to `php` file which is used like **stylesheet** . Let's see an example:
**A)** Part in `view.php` : `$_SESSION["test"]='background-color' . $atts['background_color'];`
**B)** Usage in `css`-ed `php` file : 
```
 <?php
session_start();
header("Content-type: text/css");
$backColor = $_SESSION["test"];
//$_SESSION["bg_color"];
?>

h1,h2,h3 {
  <?php echo $backColor?>;
}
```
**C)** Enqueing our style in `functions.php` : 
`wp_enqueue_style( 'theme-style-header-style', get_template_directory_uri().'/framework-customizations/extensions/shortcodes/shortcodes/section/customSectionStyle.php' );`

So , as i thought - it should work , but actually it doesn't and at this point i got stuck . 
It doesn't pass value properly in example before : 

![image](https://user-images.githubusercontent.com/24575491/34083065-be99ebf4-e372-11e7-8565-0735baf65b94.png)
BUT it works if i pass the value like this : 
`$_SESSION["test"]='background-color: aqua ;';` instead of `$_SESSION["test"]='background-color' . $atts['background_color'];`
in `view.php` 

![image](https://user-images.githubusercontent.com/24575491/34083093-621467fa-e373-11e7-8bc2-fcc8b679df66.png)

The strangest thing : that if we check the variables values with `echo` function - it gets value from array properly , but doesn't pass HEX-vaules to `php` stylesheet.

Back to `view.php`  : 
```
$_SESSION["test"]=$atts['background_color'];
echo $_SESSION["test"];
```
Result :  
![image](https://user-images.githubusercontent.com/24575491/34083142-1359fb06-e374-11e7-9679-fd34f77dfef7.png)

```
$_SESSION["test"]='background-color: aqua ;';
echo $_SESSION["test"];
```
Result : 
![image](https://user-images.githubusercontent.com/24575491/34083155-5bf1715a-e374-11e7-9963-3e50d1b0e7e7.png)

So, the main problem as i said before : it doesn't read  properly HEX- values  ,but we can easily pass color "names" like `aqua` - and i dedicated 2 hours to this issue and come with nothing

**_2) So, the second issue is : when i am enqueing `php` style in `functions.php` it overrides ALL style `style.css` , but if you can see from example above - i only change `h1-h3` styles and nothing else. 
Example of styles enqueing_**
```
wp_enqueue_style( 'theme-style', get_stylesheet_uri() );

wp_enqueue_style( 'theme-style-header-style', get_template_directory_uri().'/framework-customizations/extensions/shortcodes/shortcodes/section/customSectionStyle.php' );
 
```
**_To sum up :_**  I think that it will be better to add all styles in separate stylesheet - so we can load in cache this stylesheet olny once instead of basic logic of framework ( paragraph 1 in issues list) that requires to load section style everytime we refresh page. I am sure that my way is far not the best solution for this issue , **_although_** it may be a viable idea to avoid slow loading speed  of page if we have a lot customizable elements 
PS: sorry for my bad english - i really hope that you understood my idea .
Peace to everyone and thanks for help !

**Comment by danyj** [comment]
browse please

http://manual.unyson.io/en/latest/extension/shortcodes/#enqueue-shortcode-dynamic-css-in-page-head

https://github.com/ThemeFuse/Unyson/issues/2531

**Comment by danyj** [comment]
keyword , dynamic css

**Comment by danyj** [comment]
https://github.com/ThemeFuse/Unyson/issues/1871

**Comment by danyj** [comment]
> /customSectionStyle.php 

dont do that use , static.php in shortcode

https://github.com/ThemeFuse/Unyson-Shortcodes-Extension/blob/master/shortcodes/column/static.php

every shortcode has it for that very reason

**Comment by vivalafit** [comment]
Thanks for the reply !
I've tested your answers ,but got stuck in a little moment :
in `options.php`  i already have `custom_id`  variable and i am trying to do a thing which you have sent before (http://manual.unyson.io/en/latest/extension/shortcodes/#enqueue-shortcode-dynamic-css-in-page-head)
So , what i've achieved : 
```
wp_enqueue_style(
	'fw-shortcode-section',
	$shortcodes_extension->locate_URI( '/shortcodes/section/static/css/styles.css' )
);
//section adding style to css
if (!function_exists('_action_theme_shortcode_section_enqueue_dynamic_css')):
//start
function _action_theme_shortcode_section_enqueue_dynamic_css($data) {
    $shortcode = 'section';
    $atts = shortcode_parse_atts( $data['atts_string'] );
    $atts = fw_ext_shortcodes_decode_attr($atts, $shortcode, $data['post']->ID);
    //
    wp_add_inline_style(	
        'fw-shortcode-'.$shortcode,
        '#'. $atts['custom_id'] .' { '.
            'background-color: '. $atts['background_color'] .';'.
        ' } '
    );
}
add_action(
    'fw_ext_shortcodes_enqueue_static: section',
    '_action_theme_shortcode_section_enqueue_dynamic_css'
);
endif;
```
It actually adds such style - so i can change all elements style properly
```
#testId{
	display: none;
}
```
BUT it doesn't add any css code from shortcode above - did i missed something important ?

**Comment by vivalafit** [comment]
@danyj your method worked great - much appreciated
```
add_action(
    'fw_ext_shortcodes_enqueue_static: section',
    '_action_theme_shortcode_section_enqueue_dynamic_css'
);
endif;
```
So i missed that in  `'fw_ext_shortcodes_enqueue_static: section' ,` - there is no space between `static` and `section` 
And the right code is : `'fw_ext_shortcodes_enqueue_static:section' ,`
Thanks a lot for help

**Comment by danyj** [comment]
one more 

instead 

`$atts['background_color'] `

do 


`fw_akg('background_color',$atts,'default_value');`


advantage: no errors and default value returned if an issue plus you can get sub options like


`fw_akg('main_array/sub_option/background_color',$atts,'default_value');`

http://manual.unyson.io/en/latest/helpers/php.html#general
